sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -32,6 +32,38 @@ using cl::sycl::detail::is_sgeninteger;
 using cl::sycl::detail::queue_impl;
 using cl::sycl::detail::remove_AS;
 
+// This type trait is used to detect if the atomic operation BinaryOperation
+// used with operands of the type T is available for using in reduction.
+// The order in which the atomic operations are performed may be arbitrary and
+// thus may cause different results from run to run even on the same elements
+// and on same device. The macro SYCL_REDUCTION_DETERMINISTIC prohibits using
+// atomic operations for reduction and helps to produce stable results.
+// SYCL_REDUCTION_DETERMINISTIC is a short term solution, which perhaps become
+// deprecated eventually and is replaced by a sycl property passed to reduction.
+template <typename T, class BinaryOperation>
+using IsReduOptForFastAtomicFetch =
+#ifdef SYCL_REDUCTION_DETERMINISTIC
+    bool_constant<false>;
+#else
```

Comment:
Is this sufficient to guarantee determinism?  I don't think the work-group/sub-group reduction functions are guaranteed to be deterministic across all devices.  Since there's only one implementation that's guaranteed to be deterministic (i.e. the manual tree reduction), why not guard all but that implementation behind `ifndef SYCL_REDUCTION_DETERMINISTIC`?

Reply:
Right. I disabled reduce() algorithm when the macro is defined: ead7383168774277aaefd18e291685060cac831e